.github/workflows/ci.yml

```
@@ -194,6 +194,9 @@ jobs:
     # https://github.community/t/is-it-possible-to-require-all-github-actions-tasks-to-pass-without-enumerating-them/117957/4?u=graingert
     name: Check tag and publish
     runs-on: ubuntu-latest
+    permissions:
```

It would be good to additionally set an environment (I use `release-pypi` myself) which you then configure on PyPI to be the only one that is allowed to publish:

```suggestion
    environment: release-pypi
    permissions:
```

My understanding is that it gets created automatically when used, but you can also manually add it to https://github.com/twisted/towncrier/settings/environments

Here's a quote from the publisher docs on PyPI:

> The name of the [GitHub Actions environment](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) that the above workflow uses for publishing. This should be configured under the repository's settings. While not required, a dedicated publishing environment is strongly encouraged, especially if your repository has maintainers with commit access who shouldn't have PyPI publishing access.

---

Done. I am not sure this is needed.

I hope we can keep the towncrier access security setting simple ... and fully trust anyone with write access to the main repo.

Fork should be ok as the forks should not get access to the GitHub Secrets and I hope PyPI will deny access to the OIDC token from forks :)

---

@hynek it looks like using the GHA environment has the unwanted side effect, that any CI run is now blocked and  needs manual approval.

I tried to add a high-level conditional, but I see that the deploy request is still triggered.

---

Eh ok leave it out for now and we’ll look another time. Sorry for the noise.

---

I  think that the environment thing is worth exploring ... 

I guess that the deployment workflow is designed to be triggered as a side-effect of another workflow (with dispatch) ... and it's not designed to be executed as mixed job

---

This is how I do it: https://github.com/hynek/structlog/blob/main/.github/workflows/pypi-package.yml

If you want, you can copy it verbatim and just change the trigger conditions.

---

I think it make sense to have the release code in a separate workflow.

At the same time, I hope we can keep `towncrier` automation simple and have everything in a single file to make it easier to discover and see what is going on in GitHub Actions.

------------

I don't know what's best here :|

---

I personally think (as demonstrated :)) that CI and publishing are sufficiently distinct issues, that it absolutely makes sense to keep them apart. But that's not something we need / should solve _here_.

---

I would like to experiment with that and make another PR.